backend/src/controllers/authenticator.js

```diff
@@ -0,0 +1,67 @@
+const User = require('../mongo/data/schemas/user');
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+require('dotenv').config();
+
+exports.register = async (req, res) => {
+  const { email, password, username } = req.body;
+
+  if (!email || !password || !username) {
+    return res.status(400).json({ message: 'Email, password and username are required' });
+  }
+
+  if (!isValidEmail(email)) {
+    return res.status(400).json({ message: 'The email format is not valid' });
+  }
+
+  if (password.length < 6) {
+    return res.status(400).json({ message: 'Password must have at least 6 characters' });
+  }
+
+  try {
+    const userExists = await User.findOne({ email });
+    if (userExists) {
+      return res.status(400).json({ message: 'This email is already in use' });
+    }
+
+    const hashedPassword = await bcrypt.hash(password, 10);
+    const newUser = new User({ email, password: hashedPassword, username });
+    await newUser.save();
+
+    res.status(201).json({ message: 'User registered successfully' });
+  } catch (error) {
+    console.error('Error registering user:', error);
+    res.status(500).json({ message: 'Error registering user' });
+  }
+};
+
+const isValidEmail = (email) => {
+  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+  return emailRegex.test(email);
+};
+
+exports.login = async (req, res) => {
+  const { email, password } = req.body;
+
+  if (!email || !password) {
+    return res.status(400).json({ message: 'Email and password are required' });
+  }
+
+  try {
+    const user = await User.findOne({ email });
+    if (!user) {
+      return res.status(400).json({ message: 'User or password are incorrect' });
+    }
+
+    const isMatch = await bcrypt.compare(password, user.password);
+    if (!isMatch) {
+      return res.status(400).json({ message: 'User or password are incorrect' });
```

Comment:
No deberia ser "Email or password is incorrect" aqui? Considerando que el login es con email & password?

Reply:
Una cosa más - no sería mejor usar HTTP 401 en esta linea? https://http.cat/status/401